src/motion/cg_optimizer.F

```diff
@@ -198,14 +288,32 @@ RECURSIVE SUBROUTINE cp_cg_main(force_env, x0, gopt_param, output_unit, globenv,
          ![NB] consistent energies and forces not required for CG, but some line minimizers might set it
          save_consistent_energy_force = gopt_env%require_consistent_energy_force
          gopt_env%require_consistent_energy_force = .FALSE.
-         CALL cp_eval_at(gopt_env, x0, opt_energy, xi, gopt_env%force_env%para_env%mepos, &
-                         .FALSE., gopt_env%force_env%para_env)
+
+         ! Symmetrize coordinates and forces
+         IF (keep_sg) THEN
+            CALL cp_eval_at(gopt_env, x0, opt_energy, xi, gopt_env%force_env%para_env%mepos, &
+                            .FALSE., gopt_env%force_env%para_env, spgr=spgr)
+         ELSE
+            CALL cp_eval_at(gopt_env, x0, opt_energy, xi, gopt_env%force_env%para_env%mepos, &
+                            .FALSE., gopt_env%force_env%para_env)
+         END IF
```

Comment:
This branching is not needed. Just keep the `ELSE` block. In Fortran one can pass optional argument forward.

Reply:
That's handy and good to know.

Reply:
There are many more cases for which I do this branching than the two you commented about. Should I update all of them? For example, there is a similar case on line 135 of this file (src/motion/cg_optimizer.F):

      IF (keep_space_group) THEN
         CALL cp_cg_main(force_env, x0, gopt_param, output_unit, globenv, &
                         gopt_env, do_update=my_do_update, spgr=spgr)
      ELSE
         CALL cp_cg_main(force_env, x0, gopt_param, output_unit, globenv, &
                         gopt_env, do_update=my_do_update)
      ENDIF

Reply:
Yes, please try to remove as many of these branches as possible to keep code duplication low.

Reply:
@oschuett Alright, I removed all the branches I could find and hopefully all there are. A problem that appeared to me, though, is that I rely a lot on IF(PRESENT(spgr)) to set keep_space_group=.TRUE. Should I use the function NULL instead to check if spgr is actually defined? Is there a better option?

Reply:
Thanks. Simply write `keep_space_group = PRESENT(spgr)`. I'd consider that well-written Fortran.

The `NULL()` function only returns a null pointer in [Fortran](https://gcc.gnu.org/onlinedocs/gfortran/NULL.html).

Reply:
The thing is, without the branches, spgr is passed by default. For example:

IF (keep_space_group) THEN
   CALL cp_cg_main(force_env, x0, gopt_param, output_unit, globenv, &
                                 gopt_env, do_update=my_do_update, spgr=spgr)
ELSE
   CALL cp_cg_main(force_env, x0, gopt_param, output_unit, globenv, &
                                  gopt_env, do_update=my_do_update)
ENDIF

is now:

CALL cp_cg_main(force_env, x0, gopt_param, output_unit, globenv, &
                               gopt_env, do_update=my_do_update, spgr=spgr)

Did I misunderstand what you wanted me to do with these branches? If what I did is what you asked me to do, spgr being always there, the if(pesent(spgr)) test does not work anymore.

Reply:
No, it will still work. You can simply pass an `OPTIONAL` argument on to another function.

Reply:
@oschuett Sorry, I did not make my case clear. I have no doubts it works when I require the use of space group symmetry. I actually already checked it does when I tested that I did not delete to much or the wrong part of the branch.

My concern is for when someone is not using space group symmetry. The test present(spgr) will give true and keep_space_group will also be true. However, space group symmetry is not defined and this makes cp2k crash. I also tested that. Should I switch to if (spgr%n_operattions > 0) instead or make spgr an optional pointer rather than a variable? Would the latter affect the syntax of the functions?

Reply:
@oschuett removing the branches has unanticipated consequences. It creates many situation where spgr might be passed to routine by another routine which did not get it passed. I totally understand the need to avoid duplication of the code for the sake of maintenance, clarity, and readability, but could we leave it as it is for know and mark it as a later task. LBFGS and CP_EVAL_AT are particularly susceptible to this problem. Ideally, SPGR should be part of gopt_env. I did not do it because there are issues as well and gopt_env is not always available where I need it. I guess it all comes down to what you mention earlier: the motion code is in pretty bad shape.

Reply:
I see your point and I tend to agree that there is no simple and clean solution.

However, before we give up maybe you could, as a last attempt, try to move `spgr` into the `gopt_env`, ie. [`gopt_f_type`](https://github.com/cp2k/cp2k/blob/master/src/motion/gopt_f_types.F#L60) ?

Reply:
@oschuett I remember now one the reasons I did not do it in first place, space_group.F needs gopt_f_type. I can not call space_group module in gopt_f_type definition. An easy fix is to create a separate file where spgr type is defined: space_group_t.F. Of course, I have to check that an object with the type gopt_f_type is always passed in the routines where spgr is also passed. I will give it a try. It would certainly make the code cleaner. It would make it easier to access space group data for other use.

Reply:
Done. spgr structure is now found into gopt_f_type. I made a series of quick tests an everything seems fine to me. Hopefully, I did not forget any branches. Please let me know otherwise.